tensorboard/defs/web.bzl

```diff
@@ -14,259 +14,295 @@
+    dummy = _new_file(ctx, "-webfiles.ignoreme")
+    manifests = depset(order = "postorder", transitive = [dep.manifests for dep in deps])
+    if srcs:
+        args = [
+            "WebfilesValidator",
+            "--dummy",
+            dummy.path,
+            "--target",
+            manifest.path,
+        ]
+        if hasattr(ctx, "attr") and hasattr(ctx.attr, "suppress"):
+            for category in ctx.attr.suppress:
+                args.append("--suppress")
+                args.append(category)
+        inputs = []  # list of depsets
+        inputs.append(depset([manifest] + srcs))
+        direct_manifests = depset([dep.manifest for dep in deps])
+        for dep in deps:
+            inputs.append(depset([dep.dummy]))
+            inputs.append(depset([dep.manifest]))
+            args.append("--direct_dep")
+            args.append(dep.manifest.path)
+        for man in difference(manifests, direct_manifests):
+            inputs.append(depset([man]))
+            args.append("--transitive_dep")
+            args.append(man.path)
+        argfile = _new_file(ctx, "-webfiles-checker-args.txt")
+        ctx.actions.write(output = argfile, content = "\n".join(args))
+        inputs.append(depset([argfile]))
+        ctx.actions.run(
+            inputs = depset(transitive = inputs),
+            outputs = [dummy],
+            executable = (getattr(ctx.executable, "_ClosureWorker", None) or
+                          getattr(ctx.executable, "_ClosureWorkerAspect", None)),
+            arguments = ["@@" + argfile.path],
+            mnemonic = "Closure",
+            execution_requirements = {"supports-workers": "1"},
+            progress_message = "Checking webfiles %s" % ctx.label,
+        )
+    else:
+        ctx.actions.write(output = dummy, content = "BOO!")
+    manifests = depset([manifest], transitive = [manifests])
+    return dummy, manifests
 
 def _new_file(ctx, suffix):
-  return ctx.actions.declare_file("%s%s" % (ctx.label.name, suffix))
+    return ctx.actions.declare_file("%s%s" % (ctx.label.name, suffix))
 
 def _add_webpath(ctx, src, webpath, webpaths, new_webpaths, manifest_srcs):
-  if webpath in new_webpaths:
-    _fail(ctx, "multiple srcs within %s define the webpath %s " % (
-        ctx.label, webpath))
-  if webpath in webpaths.to_list():
-    _fail(ctx, "webpath %s was defined by %s when already defined by deps" % (
-        webpath, ctx.label))
-  new_webpaths.append(webpath)
-  manifest_srcs.append(struct(
-      path=src.path,
-      longpath=long_path(ctx, src),
-      webpath=webpath))
+    if webpath in new_webpaths:
+        _fail(ctx, "multiple srcs within %s define the webpath %s " % (
+            ctx.label,
+            webpath,
+        ))
+    if webpath in webpaths.to_list():
+        _fail(ctx, "webpath %s was defined by %s when already defined by deps" % (
+            webpath,
+            ctx.label,
+        ))
+    new_webpaths.append(webpath)
+    manifest_srcs.append(struct(
```

Comment:
Just checking... I understand this change is removing use of `struct` (is that right?)

Should I be looking to catch any uses of struct? Or it it specific ones from how they're used by rules_closure, so it's tricky to know which ones are right and which ones aren't?

Reply:
No, it's not removing all `structs`, just the structs returned directly from a rule. This one is not a problem.